### PR TITLE
Add builtin for performing checked conversion between integer types.

### PR DESCRIPTION
**Pull request by zygoloid**
As a prerequisite for switching the type of int literals to be the `IntLiteral` type, add support for performing conversions of in-bounds integer constant values to other integer types in which they fit.

This incidentally is our first compile-time-only builtin function, so add very minimal support for compile-time-only functions while we're here.

### CODE REVIEW

File: toolchain/check/testdata/builtins/int/convert_checked.carbon [hunk 1]
```diff
@@ -0,0 +1,2809 @@
...
+fn SubI32(a: i32, b: i32) -> i32 = "int.ssub";
+fn AddU32(a: u32, b: u32) -> u32 = "int.uadd";
+fn IntLiteral() -> type = "int_literal.make_type";
+
+// Size preserving
+fn Int32ToInt32(a: i32) -> i32 = "int.convert_checked";
+fn Int32ToUint32(a: i32) -> u32 = "int.convert_checked";
+fn Uint32ToInt32(a: u32) -> i32 = "int.convert_checked";
+fn Uint32ToUint32(a: u32) -> u32 = "int.convert_checked";
+fn IntLiteralToIntLiteral(a: IntLiteral()) -> IntLiteral() = "int.convert_checked";
```
Comment (josh11b) [inline toolchain/check/testdata/builtins/int/convert_checked.carbon]: Wrap?

File: toolchain/check/testdata/builtins/int/convert_checked.carbon [hunk 2]
```diff
@@ -0,0 +1,2809 @@
...
+let g: i16 = Uint32ToInt16(Int32ToUint32(0));
+let h: i16 = Uint32ToInt16(Int32ToUint32(0x7FFF));
+
+let lit_i16_min: i16 = IntLiteralToInt16(Int32ToIntLiteral(NegateI32(0x8000)));
+let lit_i16_max: i16 = IntLiteralToInt16(Int32ToIntLiteral(0x7FFF));
+
+let lit_u16_min: u16 = IntLiteralToUint16(Int32ToIntLiteral(0));
+let lit_u16_max: u16 = IntLiteralToUint16(Int32ToIntLiteral(0xFFFF));
+
+// --- zext.carbon
```
Comment (josh11b) [inline toolchain/check/testdata/builtins/int/convert_checked.carbon]:
```suggestion
// --- zero_extend.carbon
```

File: toolchain/check/testdata/builtins/int/convert_checked.carbon [hunk 3]
```diff
@@ -0,0 +1,2809 @@
...
+    AddU32(Int32ToUint32(0x7FFF_FFFF), Int32ToUint32(0x7FFF_FFFF)),
+    Int32ToUint32(1)));
+
+let c: i64 = Uint32ToInt64(Int32ToUint32(0));
+let d: i64 = Uint32ToInt64(
+  AddU32(
+    AddU32(Int32ToUint32(0x7FFF_FFFF), Int32ToUint32(0x7FFF_FFFF)),
+    Int32ToUint32(1)));
+
+// --- sext.carbon
```
Comment (josh11b) [inline toolchain/check/testdata/builtins/int/convert_checked.carbon]:
```suggestion
// --- signed_extend.carbon
```

File: toolchain/check/testdata/builtins/int/convert_checked.carbon (hunk 1 above)
Comment (zygoloid) [reply toolchain/check/testdata/builtins/int/convert_checked.carbon]: Done.

File: toolchain/check/testdata/builtins/int/convert_checked.carbon (hunk 3 above)
Comment (zygoloid) [reply toolchain/check/testdata/builtins/int/convert_checked.carbon]: Done, but `sign_extend` instead.

**Review by josh11b — APPROVED** [review]

**Pull request merged**